common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,151 @@
+use std::collections::{HashMap, VecDeque};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::Waker;
+
+use futures::async_await::FusedStream;
+use futures::stream::Stream;
+use futures::task::Context;
+use futures::Poll;
+
+/// ValidatorMessage trait needs to be implemented by every message which gets pushed into the Libra
+/// channel so that we can ensure fairness among validators
+pub trait ValidatorMessage {
+    /// Extract the Validator from which this message arrived. This
+    /// will be used for ensuring fairness among validators.
+    fn get_validator(&self) -> &str;
+}
+
+/// Type is an enum which can be used as a configuration option for
+/// libra_channel
+pub enum Type {
+    LIFO,
+    FIFO,
+}
+
+struct PerValidatorQueue<T> {
+    queue_type: Type,
+    per_validator_queue: HashMap<String, VecDeque<T>>,
+    max_queue_size: usize,
+}
+
+impl<T: ValidatorMessage> PerValidatorQueue<T> {
+    fn new(queue_type: Type, max_queue_size: usize) -> Self {
+        assert!(
+            max_queue_size > 0,
+            "cannot have a zero-capacity libra_channel"
+        );
+        Self {
+            queue_type,
+            per_validator_queue: HashMap::new(),
+            max_queue_size,
+        }
+    }
+
+    fn add_message(&mut self, message: T) {
+        let validator = message.get_validator();
+        if !self.per_validator_queue.contains_key(validator) {
+            self.per_validator_queue.insert(
+                validator.to_string(),
+                VecDeque::with_capacity(self.max_queue_size + 1),
+            );
+        }
+        if let Some(q) = self.per_validator_queue.get_mut(validator) {
+            q.push_back(message);
+            if q.len() == self.max_queue_size + 1 {
+                match self.queue_type {
+                    // Drop old message for FIFO
+                    Type::FIFO => q.pop_front(),
+                    // Drop new message for LIFO
+                    Type::LIFO => q.pop_back(),
+                };
+            }
+        }
+    }
+
+    // TODO: Come up with a better data structure to efficiently
+    // do a round robin among validators
+    fn get_message(&mut self) -> Option<T> {
+        for q in self.per_validator_queue.values_mut() {
+            if !q.is_empty() {
+                return match self.queue_type {
+                    Type::FIFO => q.pop_back(),
```

Comment:
I think it should be a reverse. The insert is using `push_back()`, so LIFO is `pop_back()` and FIFO is `pop_front()`, isn't it?

Reply:
Yup. Fixed.